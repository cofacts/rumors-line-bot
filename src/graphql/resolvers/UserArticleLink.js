export default {
  /**
   * @param {UserArticleLink} source
   */
  article(source) {
    // Returns stub
    return { id: source.articleId };
  },
};
