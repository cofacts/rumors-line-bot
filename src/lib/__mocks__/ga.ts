const eventMock = jest.fn();
const sendMock = jest.fn();
const setMock = jest.fn();

const ga = Object.assign(
  jest.fn().mockImplementation(() => ({
    event: eventMock,
    send: sendMock,
    set: setMock,
    screenview: () => undefined,
  })),
  {
    clearAllMocks: () => {
      eventMock.mockClear();
      sendMock.mockClear();
      setMock.mockClear();
      ga.mockClear();
    },
    eventMock,
    sendMock,
    setMock,
  }
);

export default ga;
