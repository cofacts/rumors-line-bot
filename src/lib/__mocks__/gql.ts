const mockResultQueue: object[] = [];

function gqlMock() {
  return () => {
    const ret = mockResultQueue.shift();
    return Promise.resolve(ret);
  };
}

/**
 * Enqueues mock data what will be returned in next gql()() call.
 *
 * Once gql()() is invoked, the first data in queue will be consumed.
 *
 * @param {*} returnValue the mock value that gql()()'s returned promise will resolve to.
 * @returns nothing
 */
gqlMock.__push = function (returnValue: object) {
  mockResultQueue.push(returnValue);
};

/**
 * Empty the return data queue.
 *
 * Invoking this in afterEach() will ensure 1 failed test would not leave side effects to other
 * tests.
 */
gqlMock.__reset = function () {
  mockResultQueue.length = 0;
};

/**
 * @returns {Boolean} if the mock response has been depleted
 */
gqlMock.__finished = function () {
  return mockResultQueue.length === 0;
};

export default gqlMock;
export type MockedGql = typeof gqlMock;
