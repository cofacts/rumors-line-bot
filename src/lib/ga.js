import ua from 'universal-analytics';
import { gaTitle } from './sharedUtils';

/**
 * Sends a screen view and returns the visitor
 *
 * @param {string} uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param {string} state - The current state the LINE bot is in
 * @param {string} documentTitle - The article content (user input) currently in search
 * @param {string} messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns {ua.Visitor} universal analytics visitor
 */
export default function ga(
  uuid,
  state = 'N/A',
  documentTitle = '',
  messageSource = 'user'
) {
  const visitor = ua(process.env.GA_ID, uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  return visitor;
}
