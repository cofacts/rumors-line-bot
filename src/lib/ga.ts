import ua from 'universal-analytics';
import type { EventParams } from 'universal-analytics';
import { gaTitle } from './sharedUtils';
import { insertEventBatch } from './bq';
import type { EventBatch } from 'src/rumors-db/bq/events';
import rollbar from './rollbar';

/**
 * Sends a screen view and returns the visitor
 *
 * @param uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param state - The current state the LINE bot is in
 * @param documentTitle - The article content (user input) currently in search
 * @param messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns universal analytics visitor
 */
export default function ga(
  uuid: string,
  state = 'N/A',
  documentTitle = '',
  messageSource: 'user' | 'group' | 'room' = 'user'
) {
  const visitor = ua(process.env.GA_ID || '', uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  let events: EventBatch['events'] = [];
  const extra: Record<string, unknown> = {};

  return {
    set(key: string, value: unknown) {
      extra[key] = value;
      return visitor.set(key, value);
    },

    event(evt: EventParams) {
      events.push({
        category: evt.ec ?? null,
        action: evt.ea ?? null,
        label: evt.el ?? null,
        value: evt.ev === undefined ? null : +evt.ev,
        time: new Date(),
      });
      return visitor.event(evt);
    },
    send() {
      insertEventBatch({
        text: documentTitle,
        messageSource,
        events,
        createdAt: new Date(),
        extra,
      }).catch((e) => {
        console.error('[insertAnalytics]', e);
        rollbar.error(`[insertAnalytics] ${e.message}`, e);
      });
      events = [];
      return visitor.send();
    },
  };
}
