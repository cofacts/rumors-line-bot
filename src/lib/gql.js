import fetch from 'node-fetch';
import rollbar from './rollbar';
import url from 'url';

const API_URL =
  process.env.API_URL || 'https://cofacts-api.hacktabl.org/graphql';

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
// GraphQL Protocol: http://dev.apollodata.com/tools/graphql-server/requests.html
//
export default (query, ...substitutions) => (variables, search) => {
  const queryAndVariable = {
    query: String.raw(query, ...substitutions),
  };

  if (variables) queryAndVariable.variables = variables;

  let status;
  const URL = `${API_URL}${url.format({ query: search })}`;

  return fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-app-secret': process.env.APP_SECRET,
    },
    credentials: 'include',
    body: JSON.stringify(queryAndVariable),
  })
    .then(r => {
      status = r.status;
      return r.json();
    })
    .then(resp => {
      if (status === 400) {
        throw new Error(
          `GraphQL Error: ${resp.errors
            .map(({ message }) => message)
            .join('\n')}`
        );
      }
      if (resp.errors) {
        // When status is 200 but have error, just print them out.
        console.error('GraphQL operation contains error:', resp.errors);
        rollbar.error(
          'GraphQL error',
          {
            body: JSON.stringify(queryAndVariable),
            url: URL,
          },
          { resp }
        );
      }
      return resp;
    });
};
