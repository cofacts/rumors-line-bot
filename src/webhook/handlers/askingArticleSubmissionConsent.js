import { t } from 'ttag';
import ga from 'src/lib/ga';
import gql from 'src/lib/gql';
import { SOURCE_PREFIX, getArticleURL } from 'src/lib/sharedUtils';
import {
  ManipulationError,
  createArticleShareBubble,
  createSuggestOtherFactCheckerReply,
  getArticleSourceOptionFromLabel,
  createReasonButtonFooter,
} from './utils';

import UserArticleLink from '../../database/models/userArticleLink';

export default async function askingArticleSubmissionConsent(params) {
  let { data, state, event, userId, replies, isSkipUser } = params;

  if (!event.input.startsWith(SOURCE_PREFIX)) {
    throw new ManipulationError(
      t`Please press the latest button to submit message to database.`
    );
  }

  const visitor = ga(userId, state, data.searchedText);

  const sourceOption = getArticleSourceOptionFromLabel(
    event.input.slice(SOURCE_PREFIX.length)
  );

  visitor.event({
    ec: 'UserInput',
    ea: 'ProvidingSource',
    el: sourceOption.value,
  });

  if (!sourceOption.valid) {
    replies = [
      {
        type: 'text',
        text: t`Thanks for the info.`,
      },
      createSuggestOtherFactCheckerReply(),
    ];
    state = '__INIT__';
  } else {
    visitor.event({ ec: 'Article', ea: 'Create', el: 'Yes' });

    const {
      data: { CreateArticle },
    } = await gql`
      mutation($text: String!) {
        CreateArticle(text: $text, reference: { type: LINE }) {
          id
        }
      }
    `({ text: data.searchedText }, { userId });

    await UserArticleLink.create({ userId, articleId: CreateArticle.id });

    const articleUrl = getArticleURL(CreateArticle.id);
    const articleCreatedMsg = t`Your submission is now recorded at ${articleUrl}`;

    // Track the source of the new message.
    visitor.event({
      ec: 'Article',
      ea: 'ProvidingSource',
      el: `${CreateArticle.id}/${sourceOption.value}`,
    });

    replies = [
      {
        type: 'flex',
        altText: articleCreatedMsg,
        contents: {
          type: 'carousel',
          contents: [
            {
              type: 'bubble',
              body: {
                type: 'box',
                layout: 'vertical',
                contents: [
                  {
                    type: 'text',
                    wrap: true,
                    text: articleCreatedMsg,
                  },
                ],
              },
              footer: createReasonButtonFooter(
                articleUrl,
                userId,
                data.sessionId
              ),
            },
            createArticleShareBubble(articleUrl),
          ],
        },
      },
    ];

    // Record article ID in context for reason LIFF
    data.selectedArticleId = CreateArticle.id;
    state = '__INIT__';
  }

  visitor.send();
  return { data, state, event, userId, replies, isSkipUser };
}
