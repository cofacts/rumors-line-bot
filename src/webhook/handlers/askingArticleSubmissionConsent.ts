import { t } from 'ttag';
import { Message } from '@line/bot-sdk';
import { z } from 'zod';

import { ChatbotPostbackHandler } from 'src/types/chatbotState';
import ga from 'src/lib/ga';
import gql from 'src/lib/gql';
import { getArticleURL } from 'src/lib/sharedUtils';
import UserSettings from 'src/database/models/userSettings';
import UserArticleLink from 'src/database/models/userArticleLink';
import {
  ArticleTypeEnum,
  SubmitMediaArticleUnderConsentMutation,
  SubmitMediaArticleUnderConsentMutationVariables,
  SubmitTextArticleUnderConsentMutation,
  SubmitTextArticleUnderConsentMutationVariables,
} from 'typegen/graphql';

import {
  POSTBACK_YES,
  POSTBACK_NO,
  ManipulationError,
  createTextMessage,
  createCommentBubble,
  createArticleShareBubble,
  createNotificationSettingsBubble,
  getLineContentProxyURL,
  createAIReply,
} from './utils';

const inputSchema = z.enum([POSTBACK_NO, POSTBACK_YES]);

/** Postback input type for ASKING_ARTICLE_SUBMISSION_CONSENT state handler */
export type Input = z.infer<typeof inputSchema>;

function uppercase<T extends string>(s: T) {
  return s.toUpperCase() as Uppercase<T>;
}

const askingArticleSubmissionConsent: ChatbotPostbackHandler = async ({
  context,
  postbackData: { state, input: postbackInput },
  userId,
}) => {
  let input: Input;
  try {
    input = inputSchema.parse(postbackInput);
  } catch (e) {
    console.error('[askingArticleSubmissionConsnet]', e);
    throw new ManipulationError(t`Please choose from provided options.`);
  }

  const firstMsg = context.msgs[0];
  // istanbul ignore if
  if (!firstMsg) {
    throw new ManipulationError('No message found in context'); // Should never happen
  }

  const visitor = ga(
    userId,
    state,
    firstMsg.type === 'text' ? firstMsg.text : firstMsg.id
  );

  let replies: Message[] = [];

  switch (input) {
    default:
      // Exhaustive check
      return input satisfies never;

    case POSTBACK_NO:
      visitor.event({ ec: 'Article', ea: 'Create', el: 'No' });
      replies = [
        createTextMessage({
          text: t`The message has not been reported and won’t be fact-checked. Thanks anyway!`,
        }),
      ];
      break;

    case POSTBACK_YES: {
      visitor.event({ ec: 'Article', ea: 'Create', el: 'Yes' });
      let article;
      if (firstMsg.type === 'text') {
        const result = await gql`
          mutation SubmitTextArticleUnderConsent($text: String!) {
            CreateArticle(text: $text, reference: { type: LINE }) {
              id
            }
          }
        `<
          SubmitTextArticleUnderConsentMutation,
          SubmitTextArticleUnderConsentMutationVariables
        >({ text: firstMsg.text }, { userId });
        article = result.data.CreateArticle;
      } else {
        const articleType: ArticleTypeEnum = uppercase(firstMsg.type);

        const proxyUrl = getLineContentProxyURL(firstMsg.id);

        const result = await gql`
          mutation SubmitMediaArticleUnderConsent(
            $mediaUrl: String!
            $articleType: ArticleTypeEnum!
          ) {
            CreateMediaArticle(
              mediaUrl: $mediaUrl
              articleType: $articleType
              reference: { type: LINE }
            ) {
              id
            }
          }
        `<
          SubmitMediaArticleUnderConsentMutation,
          SubmitMediaArticleUnderConsentMutationVariables
        >({ mediaUrl: proxyUrl, articleType }, { userId });
        article = result.data.CreateMediaArticle;
      }

      /* istanbul ignore if */
      if (!article?.id) {
        throw new Error(
          '[askingArticleSubmissionConsent] article is not created successfully'
        );
      }

      await UserArticleLink.createOrUpdateByUserIdAndArticleId(
        userId,
        article.id
      );

      // Create new session, make article submission button expire after submission
      context.sessionId = Date.now();

      const articleUrl = getArticleURL(article.id);
      const articleCreatedMsg = t`Your submission is now recorded at ${articleUrl}`;
      const { allowNewReplyUpdate } = await UserSettings.findOrInsertByUserId(
        userId
      );

      let maybeAIReplies: Message[] = [
        createTextMessage({
          text: t`In the meantime, you can:`,
        }),
      ];

      if (firstMsg.type === 'text') {
        const aiReply = await createAIReply(article.id, userId);

        if (aiReply) {
          maybeAIReplies = [
            createTextMessage({
              text: '這篇文章尚待查核中，請先不要相信這篇文章。\n以下是機器人初步分析此篇訊息的結果，希望能帶給你一些想法。',
            }),
            aiReply,
            createTextMessage({
              text: '讀完以上機器人的自動分析後，您可以：',
            }),
          ];
        }
      }

      replies = [
        {
          type: 'flex',
          altText: t`The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!`,
          contents: {
            type: 'bubble',
            body: {
              type: 'box',
              layout: 'vertical',
              contents: [
                {
                  type: 'text',
                  wrap: true,
                  text: t`The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!`,
                },
                {
                  type: 'button',
                  action: {
                    type: 'uri',
                    label: t`View reported message`,
                    uri: articleUrl,
                  },
                  margin: 'md',
                },
              ],
            },
          },
        },
        ...maybeAIReplies,
        {
          type: 'flex',
          altText: articleCreatedMsg,
          contents: {
            type: 'carousel',
            contents: [
              createCommentBubble(article.id),
              // Ask user to turn on notification if the user did not turn it on
              //
              process.env.NOTIFY_METHOD &&
                !allowNewReplyUpdate &&
                createNotificationSettingsBubble(),
              createArticleShareBubble(articleUrl),
            ].filter(Boolean),
          },
        },
      ];
    }
  }

  visitor.send();
  return { context, replies };
};

export default askingArticleSubmissionConsent;
