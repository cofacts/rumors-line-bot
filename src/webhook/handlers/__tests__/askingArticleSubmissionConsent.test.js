jest.mock('src/lib/gql');
jest.mock('src/lib/ga');

import MockDate from 'mockdate';
import askingArticleSubmissionConsent from '../askingArticleSubmissionConsent';
import {
  REASON_PREFIX,
  SOURCE_PREFIX,
  ARTICLE_SOURCE_OPTIONS,
} from 'src/lib/sharedUtils';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';

import UserArticleLink from '../../../database/models/userArticleLink';
import Client from '../../../database/mongoClient';

beforeAll(async () => {
  if (await UserArticleLink.collectionExists()) {
    await (await UserArticleLink.client).drop();
  }
});

afterAll(async () => {
  await (await Client.getInstance()).close();
});

beforeEach(() => {
  ga.clearAllMocks();
});

it('should block incorrect prefix', async () => {
  const params = {
    data: {
      searchedText: 'Some text forwarded by the user',
      foundArticleIds: [],
    },
    state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'foo', // Wrong prefix
    },
  };

  expect(askingArticleSubmissionConsent(params)).rejects.toMatchInlineSnapshot(
    `[Error: Please press the latest button to submit message to database.]`
  );
});

it('should block non-existence source option', async () => {
  const params = {
    data: {
      searchedText: 'Some text forwarded by the user',
      foundArticleIds: [],
    },
    state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    event: {
      type: 'message',
      input: SOURCE_PREFIX + 'foo', // Correct prefix with wrong value
    },
  };

  await expect(
    askingArticleSubmissionConsent(params)
  ).rejects.toMatchInlineSnapshot(
    `[Error: Please tell us where you have received the message using the options we provided.]`
  );
});

it('should redirect user to other fact-checkers for invalid options', async () => {
  const params = {
    data: {
      searchedText: 'Some text forwarded by the user',
      foundArticleIds: [],
    },
    state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    event: {
      type: 'message',
      input:
        SOURCE_PREFIX +
        ARTICLE_SOURCE_OPTIONS.find(({ valid }) => !valid).label, // Correct prefix + option that should not proceed
    },
  };

  const result = await askingArticleSubmissionConsent(params);
  expect(result).toMatchSnapshot();
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "ProvidingSource",
          "ec": "UserInput",
          "el": "outside LINE",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('should submit article when valid source is provided', async () => {
  const params = {
    data: {
      searchedText: 'Some text forwarded by the user',
      foundArticleIds: [],
    },
    state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    event: {
      type: 'message',
      input:
        SOURCE_PREFIX + ARTICLE_SOURCE_OPTIONS.find(({ valid }) => valid).label,
    },
    userId: 'userId',
  };

  MockDate.set('2020-01-01');
  gql.__push({ data: { CreateArticle: { id: 'new-article-id' } } });
  const result = await askingArticleSubmissionConsent(params);
  MockDate.reset();
  expect(gql.__finished()).toBe(true);

  expect(result).toMatchSnapshot();
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "ProvidingSource",
          "ec": "UserInput",
          "el": "group message",
        },
      ],
      Array [
        Object {
          "ea": "Create",
          "ec": "Article",
          "el": "Yes",
        },
      ],
      Array [
        Object {
          "ea": "ProvidingSource",
          "ec": "Article",
          "el": "new-article-id/group message",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('should create a UserArticleLink when creating a Article', async () => {
  const userId = 'user-id-0';
  const params = {
    data: {
      searchedText: 'Some text forwarded by the user',
      foundArticleIds: [],
    },
    state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    event: {
      type: 'message',
      input:
        SOURCE_PREFIX + ARTICLE_SOURCE_OPTIONS.find(({ valid }) => valid).label,
    },
    userId,
  };

  MockDate.set('2020-01-01');
  gql.__push({ data: { CreateArticle: { id: 'new-article-id' } } });
  await askingArticleSubmissionConsent(params);
  MockDate.reset();

  await UserArticleLink.findByUserId(userId);
  const userArticleLinks = await UserArticleLink.findByUserId(userId);
  expect(userArticleLinks.map(e => ({ ...e, _id: '_id' }))).toMatchSnapshot();
});
