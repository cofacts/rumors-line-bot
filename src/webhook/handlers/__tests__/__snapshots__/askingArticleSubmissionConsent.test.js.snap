// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a UserArticleLink when creating a Article 1`] = `
Array [
  Object {
    "_id": "_id",
    "articleId": "new-article-id",
    "createdAt": 2020-01-01T00:00:00.000Z,
    "userId": "user-id-0",
  },
]
`;

exports[`should redirect user to other fact-checkers for invalid options 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [],
    "searchedText": "Some text forwarded by the user",
  },
  "event": Object {
    "input": "‚ÑπÔ∏è I got the message from:
Somewhere outside LINE",
    "type": "message",
  },
  "isSkipUser": undefined,
  "replies": Array [
    Object {
      "text": "Thanks for the info.",
      "type": "text",
    },
    Object {
      "altText": "We suggest forwarding the message to the following fact-checkers instead. They have üíÅ 1-on-1 Q&A service to respond to your questions.

üì± Please proceed on your mobile phone.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "We suggest forwarding the message to the following fact-checkers instead. They have üíÅ 1-on-1 Q&A service to respond to your questions.",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "MyGoPen Ë≥£Êì±È®ô",
                "type": "uri",
                "uri": "line://ti/p/%40mygopen",
              },
              "color": "#333333",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "label": "Ëò≠ÂßÜÈÖíÂêêÂè∏",
                "type": "uri",
                "uri": "line://ti/p/%40rumtoast",
              },
              "color": "#333333",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "spacing": "sm",
          "type": "box",
        },
        "styles": Object {
          "body": Object {
            "separator": true,
          },
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
  "state": "__INIT__",
  "userId": undefined,
}
`;

exports[`should submit article when valid source is provided 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [],
    "searchedText": "Some text forwarded by the user",
    "selectedArticleId": "new-article-id",
  },
  "event": Object {
    "input": "‚ÑπÔ∏è I got the message from:
A LINE group",
    "type": "message",
  },
  "isSkipUser": undefined,
  "replies": Array [
    Object {
      "altText": "Your submission is now recorded at https://cofacts.hacktabl.org/article/new-article-id",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "Your submission is now recorded at https://cofacts.hacktabl.org/article/new-article-id",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "See reported message",
                    "type": "uri",
                    "uri": "https://cofacts.hacktabl.org/article/new-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "Provide more info",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW/liff/index.html?p=reason&token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1c2VySWQiLCJleHAiOjE1Nzc5MjMyMDB9.4tsRDmoKzX7b9gex8UyEgo2ht9hCDAQ_45CTgJ7G8Fg",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "Your friends may know the answer üåü Share your question to friends, maybe someone can help!",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Share on LINE",
                    "type": "uri",
                    "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "Share on Facebook",
                    "type": "uri",
                    "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
  "state": "__INIT__",
  "userId": "userId",
}
`;
